include/nbl/builtin/hlsl/bxdf/common.hlsl

```diff
@@ -48,358 +139,976 @@ struct Basic
+        const bool transmitted, NBL_CONST_REF_ARG(vector3_type) V, NBL_CONST_REF_ARG(vector3_type) L,
+        NBL_CONST_REF_ARG(vector3_type) T, NBL_CONST_REF_ARG(vector3_type) B, NBL_CONST_REF_ARG(vector3_type) N,
+        const scalar_type NdotL, const scalar_type VdotL,
+        const scalar_type orientedEta, const scalar_type rcpOrientedEta, NBL_REF_ARG(vector3_type) H
+    )
+    {
+        const bool valid = isocache_type::compute(retval.iso_cache,transmitted,V,L,N,NdotL,VdotL,orientedEta,rcpOrientedEta,H);
+        if (valid)
+        {
+            retval.TdotH = nbl::hlsl::dot<vector3_type>(T,H);
+            retval.BdotH = nbl::hlsl::dot<vector3_type>(B,H);
+        }
+        return valid;
+    }
+    template<class ObserverRayDirInfo, class IncomingRayDirInfo>
+    static bool compute(
+        NBL_REF_ARG(this_t) retval,
+        NBL_CONST_REF_ARG(surface_interactions::SAnisotropic<ObserverRayDirInfo>) interaction,
+        NBL_CONST_REF_ARG(SLightSample<IncomingRayDirInfo>) _sample,
+        const scalar_type eta
+    )
+    {
+        vector3_type H;
+        const bool valid = isocache_type::template compute<ObserverRayDirInfo, IncomingRayDirInfo>(retval.iso_cache,interaction.isotropic,_sample,eta,H);
+        if (valid)
+        {
+            retval.TdotH = nbl::hlsl::dot<vector3_type>(interaction.getT(),H);
+            retval.BdotH = nbl::hlsl::dot<vector3_type>(interaction.getB(),H);
+        }
+        return valid;
+    }
+
+    scalar_type getTdotH() NBL_CONST_MEMBER_FUNC { return TdotH; }
+    scalar_type getBdotH() NBL_CONST_MEMBER_FUNC { return BdotH; }
+
+    isocache_type iso_cache;
+    scalar_type TdotH;
+    scalar_type BdotH;
+};
+
+template<typename CacheT>
+struct microfacet_cache_traits;
+
+template<typename U>
+struct microfacet_cache_traits<SIsotropicMicrofacetCache<U> >
+{
+    NBL_CONSTEXPR_STATIC_INLINE bool is_aniso = false;
+};
+
+template<typename U>
+struct microfacet_cache_traits<SAnisotropicMicrofacetCache<U> >
+{
+    NBL_CONSTEXPR_STATIC_INLINE bool is_aniso = true;
+};
+
+
+#define NBL_CONCEPT_NAME generalized_spectral_of
+#define NBL_CONCEPT_TPLT_PRM_KINDS (typename)(typename)
+#define NBL_CONCEPT_TPLT_PRM_NAMES (T)(F)
+#define NBL_CONCEPT_PARAM_0 (spec, T)
+#define NBL_CONCEPT_PARAM_1 (field, F)
+NBL_CONCEPT_BEGIN(2)
+#define spec NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_0
+#define field NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_1
+NBL_CONCEPT_END(
+    //((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((spec[field]), ::nbl::hlsl::is_scalar_v))  // correctness?
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((spec * field), ::nbl::hlsl::is_same_v, T))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((field * spec), ::nbl::hlsl::is_same_v, T))
+    ((NBL_CONCEPT_REQ_TYPE_ALIAS_CONCEPT)(is_scalar_v, F))
+);
+#undef field
+#undef spec
+#include <nbl/builtin/hlsl/concepts/__end.hlsl>
+
+template<typename T, typename F>
+NBL_BOOL_CONCEPT spectral_of = generalized_spectral_of<T,F> || is_vector_v<T> || is_scalar_v<T>;
```

Comment:
just fold everything into the `generalized_spectral_of` concept, no need for extra template variable